File: LC/python/Reorder List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return nothing
    def reorderList(self, head):
        if head == None or head.next == None:
            return head
        
        #slice list to two
        slow = head
        quick = head
        while quick.next and quick.next.next:
            slow = slow.next
            quick = quick.next.next
            
        head1 = head
        head2 = slow.next
        slow.next = None
        
        #reverse list2
        p1 = head2
        p2 = head2.next
        head2.next = None
        while p2:
            tmp = p2
            p2 = p2.next
            tmp.next = p1
            p1 = tmp
        head2 = p1
        
        #merge two lists
        pf = head1
        pe = head2
        while pe:
            temp1 = pf.next
            temp2 = pe.next
            pf.next = pe
            pe.next = temp1
            pf = temp1
            pe = temp2
        
        
            

File: LC/java/Reorder List.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode reorderList(ListNode head) {
        if(head == null || head.next == null){
            return head;
        }
        
        ListNode res,slow,fast,head2,a,b,tmp,r,p1,p2;
        res = head;
        slow = head;
        fast = head;
        
        while(fast.next!=null && fast.next.next != null){
            fast = fast.next.next;
            slow = slow.next;
        }
        head2 = slow.next;
        slow.next = null;
        
        //reverse the second part
        a = head2;
        b = head2.next;
        head2.next = null;
        while(b!=null){
            tmp = b;
            b = b.next;
            tmp.next = a;
            a = tmp;
        }
        head2 = a;
     
        //merge
        while(head!=null && head2!=null){
            p1 = head.next;
            head.next = head2;
            p2 = head2.next;
            head2.next = p1;
            head = p1;
            head2 = p2;
        }
        
        if (head != null){
            head.next = null;
        }
        return res;
    }
}